### PR TITLE
Allow H1, H2, and H2C protocol for HttpClient

### PR DESCRIPTION
**Pull request by alextychan**
Currently a client that's configured for HTTP 1.1, H2, and H2C is unable to make requests to both HTTPS and HTTP urls due to checks in HttpClientConnect.java which can be safely ignored as long as a backup protocol is configured.

The following changes are in line with discussion in this PR [#1403](https://github.com/reactor/reactor-netty/pull/1403#issuecomment-737087183). Specifically the table as follows.

Configured Client Protocols | Server URI scheme | Used Protocol | Current Status
--- | --- | --- | ---
{`H2`} | `http` | _error_ | Works
{`H2`} | `https` | `H2` (pre-known) | Works
{`H2C`} | `http` | `H2C` (pre-known) | Works
{`H2C`} | `https` | _error_ | Works
{`HTTP11`} | `http` | `HTTP11` | Works
{`HTTP11`} | `https` | `HTTP11` | Works
{`H2`, `H2C`} | `http` | `H2C` (pre-known) | Works
{`H2`, `H2C`} | `https` | `H2` (pre-known) | Works
{`H2`, `HTTP11`} | `http` | `HTTP11` | Works
{`H2`, `HTTP11`} | `https` | `H2` (offered via ALPN, fallback to `HTTP11`) | Works
{`H2C`, `HTTP11`} | `http` | `H2C` (offered via `Upgrade` header, fallback to `HTTP11`) | Works
{`H2C`, `HTTP11`} | `https` | `HTTP11` | Works
{`H2`, `H2C`, `HTTP11`} | `http` | `H2C` (offered via `Upgrade` header, fallback to `HTTP11`).  | Fails with "Configured H2 protocol without TLS"
{`H2`, `H2C`, `HTTP11`} | `https` | `H2` (offered via ALPN, fallback to `HTTP11`) | Fails with "Configured H2 Clear-Text protocol with TLS"

This fix will allow clients configured with HTTP 1.1, H2, and H2C to work as expected by ignoring H2C in HTTPS, and by ignoring H2 in HTTP.

### CODE REVIEW

File: reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConnect.java [hunk 1]
```diff
@@ -214,13 +214,13 @@ public void subscribe(CoreSubscriber<? super Connection> actual) {
...
-							removeIncompatibleProtocol(_config, HttpProtocol.H2C);
-						}
 						_config.sslProvider = HttpClientSecure.defaultSslProvider(_config);
 					}
 
-					if (_config.checkProtocol(HttpClientConfig.h2c)) {
+					if (_config.checkProtocol(HttpClientConfig.h2c) && _config.protocols.length > 1) {
+						removeIncompatibleProtocol(_config, HttpProtocol.H2C);
+					}
+					else if (_config.checkProtocol(HttpClientConfig.h2c)) {
```
Comment (violetagg) [inline reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConnect.java:223]: What do you think about checking once for `_config.checkProtocol(HttpClientConfig.h2c)`?

File: reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConnect.java [hunk 2]
```diff
@@ -242,13 +242,13 @@ public void subscribe(CoreSubscriber<? super Connection> actual) {
...
-							removeIncompatibleProtocol(_config, HttpProtocol.H2);
-						}
 						_config.sslProvider = null;
 					}
 
-					if (_config.checkProtocol(HttpClientConfig.h2)) {
+					if (_config.checkProtocol(HttpClientConfig.h2) && _config.protocols.length > 1) {
+						removeIncompatibleProtocol(_config, HttpProtocol.H2);
+					}
+					else if (_config.checkProtocol(HttpClientConfig.h2)) {
```
Comment (violetagg) [inline reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConnect.java:251]: What do you think about checking once for `_config.checkProtocol(HttpClientConfig.h2)`?

**Review by violetagg — CHANGES_REQUESTED** [review]
Can you provide also test that demonstrates how HTTP and HTTPS urls will behave when the client is configured with all 3 types of protocol?

File: reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConnect.java (hunk 1 above)
Comment (alextychan) [reply reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConnect.java]: Sounds good.

File: reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConnect.java (hunk 2 above)
Comment (alextychan) [reply reactor-netty-http/src/main/java/reactor/netty/http/client/HttpClientConnect.java]: Sounds good.

**Comment by alextychan** [comment]
I've only added tests for edge cases where the client is configured with all 3 types of protocol for the following cases.

1. Calling a https url with ssl configured. 
    * Previously: Fails with "Configured H2 Clear-Text protocol with TLS"
    * Now: Should succeed by ignoring H2C protocol
2. Calling a http url without ssl configured.
    * Previously: Fails with "Configured H2 protocol without TLS"
    * Now: Should succeed by ignoring H2 protocol

**Review by violetagg — APPROVED** [review]
@alextychan Thanks for the PR

@reactor/netty-team PTAL

**Comment by violetagg** [comment]
The failing QUIC tests on CI are not related

**Review by pderop — APPROVED** [review]
looking good to me, 

great patch; thanks @alextychan !
